vault/external_plugin_test.go

```diff
@@ -0,0 +1,404 @@
+package vault
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/hashicorp/vault/helper/namespace"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+)
+
+var (
+	compileAuthOnce   sync.Once
+	compileSecretOnce sync.Once
+	pluginBytes       []byte
```

Comment:
It looks like this needs to be split out into separate byte slices for the two different plugins, otherwise they cache the wrong type of plugin.

Something like this appears to work:

```diff
diff --git a/vault/external_plugin_test.go b/vault/external_plugin_test.go
index 00ea3c4114..a735282219 100644
--- a/vault/external_plugin_test.go
+++ b/vault/external_plugin_test.go
@@ -21,7 +21,8 @@ import (
 var (
 	compileAuthOnce   sync.Once
 	compileSecretOnce sync.Once
-	pluginBytes       []byte
+	secretPluginBytes []byte
+	authPluginBytes   []byte
 )
 
 func testCoreWithPlugin(t *testing.T, typ consts.PluginType) (*Core, string, string) {
@@ -42,17 +43,20 @@ func compilePlugin(t *testing.T, typ consts.PluginType) (name string, shasum str
 
 	var pluginType, pluginName, builtinDirectory string
 	var once *sync.Once
+	var pluginBytes *[]byte
 	switch typ {
 	case consts.PluginTypeCredential:
 		pluginType = "approle"
 		pluginName = "vault-plugin-auth-" + pluginType
 		builtinDirectory = "credential"
 		once = &compileAuthOnce
+		pluginBytes = &authPluginBytes
 	case consts.PluginTypeSecrets:
 		pluginType = "consul"
 		pluginName = "vault-plugin-secrets-" + pluginType
 		builtinDirectory = "logical"
 		once = &compileSecretOnce
+		pluginBytes = &secretPluginBytes
 	default:
 		t.Fatal(typ.String())
 	}
@@ -80,23 +84,24 @@ func compilePlugin(t *testing.T, typ consts.PluginType) (name string, shasum str
 		if err != nil {
 			t.Fatal(fmt.Errorf("error running go build %v output: %s", err, output))
 		}
-		pluginBytes, err = os.ReadFile(pluginPath)
+		actualPluginBytes, err := os.ReadFile(pluginPath)
 		if err != nil {
 			t.Fatal(err)
 		}
+		*pluginBytes = actualPluginBytes
 	})
 
 	// write the cached plugin if necessary
 	var err error
 	if _, err := os.Stat(pluginPath); os.IsNotExist(err) {
-		err = os.WriteFile(pluginPath, pluginBytes, 0o777)
+		err = os.WriteFile(pluginPath, *pluginBytes, 0o777)
 	}
 	if err != nil {
 		t.Fatal(err)
 	}
 
 	sha := sha256.New()
-	_, err = sha.Write(pluginBytes)
+	_, err = sha.Write(*pluginBytes)
 	if err != nil {
 		t.Fatal(err)
 	}
```

Reply:
Thanks!